_blog/2024-02-08-uptime-monitoring-rust.mdx

```diff
@@ -0,0 +1,694 @@
+
+```rust
+fn fill_data_gaps(
+    mut data: Vec<WebsiteStats>,
+    splits: i32,
+    format: SplitBy,
+    number_of_seconds: i32
+) -> Vec<WebsiteStats> {
+    // if the length of data is not as long as the number of required splits (24)
+    // then we fill in the gaps
+    if (data.len() as i32) < splits {
+        // for each split, format the time and check if the timestamp exists
+        for i in 1..24 {
+            let time = Utc::now() - chrono::Duration::seconds((number_of_seconds * i).into());
+            let time = time
+                .with_minute(0)
+                .unwrap()
+                .with_second(0)
+                .unwrap()
+                .with_nanosecond(0)
+                .unwrap();
+
+            let time = if matches!(format, SplitBy::Day) {
+                time.with_hour(0).unwrap()
+            } else {
+                time
+            };
+            // if timestamp doesn't exist, push a timestamp with None
+            if !data.iter().any(|x| x.time == time) {
+                data.push(WebsiteStats {
+                    time,
+                    uptime_pct: None,
+                });
+            }
+        }
+        // finally, sort the data
+        data.sort_by(|a, b| b.time.cmp(&a.time));
+    }
+
+    data
+}
+```
+
+Once done, we’ll want to create a dynamic route for getting more information about a monitored URL. We can use this page to display things like past incidents/alerts. This function will follow most of the previous handler function except we’re fetching one URL and additionally grabbing any records of HTTP requests that didn’t return `200 OK` and labelling them as “Incidents”. 
+
+As you can see below it is *mostly* the same as grabbing data for all websites, except now we also have last month’s data (split by day) to add to our return results:
+
+```rust
+#[derive(Serialize, sqlx::FromRow, Template)]
+#[template(path = "single_website.html")]
+struct SingleWebsiteLogs {
+    log: WebsiteInfo,
+    incidents: Vec<Incident>,
+    monthly_data: Vec<WebsiteStats>,
+}
+
+#[derive(sqlx::FromRow, Serialize)]
+pub struct Incident {
+    time: DateTime<Utc>,
+    status: i16,
+}
+
+async fn get_website_by_alias(
+    State(state): State<AppState>,
+    Path(alias): Path<String>,
+) -> Result<impl AskamaIntoResponse, ApiError> {
+    let website = sqlx::query_as::<_, Website>("SELECT url, alias FROM websites WHERE alias = $1")
+        .bind(&alias)
+        .fetch_one(&state.db)
+        .await?;
+
+    let last_24_hours_data = get_daily_stats(&website.alias, &state.db).await?;
+    let monthly_data = get_monthly_stats(&website.alias, &state.db).await?;
+
+    let incidents = sqlx::query_as::<_, Incident>(
+        "SELECT logs.created_at as time, 
+         logs.status from logs 
+         left join websites on websites.id = logs.website_id 
+         where websites.alias = $1 and logs.status != 200",
+    )
+    .bind(&alias)
+    .fetch_all(&state.db)
+    .await?;
```

Comment:
Since the first two functions are factored out, it would be nice to factor out this DB query too.